source/apiVolontaria/volunteer/admin.py

```diff
@@ -88,7 +92,88 @@ class EventAdmin(admin.ModelAdmin):
     ordering = ('start_date',)
 
 
-admin.site.register(models.Cycle)
+def cycle_admin_generate_participation_report(modeladmin, request, queryset):
+    if len(queryset) != 1:
+        messages.error(request, _("You must select a cycle"))
+        return
+
+    # Take the first of the array
+    cycle = queryset[0]
+
+    # Select all the Event of this Cycle
+    events_in_cycle = models.Event.objects.filter(cycle=cycle)
+
+    # Select all the participation of this Cycle
+    participations = models.Participation.objects.filter(
+        event__in=events_in_cycle
+    )
+
+    data_dict = {}
+
+    for obj in participations:
+        # We don't want to generate report
+        # if some Participation are not initialised...
+        # There is a message in the admin for this case
+        if obj.presence_status == models.Participation.PRESENCE_STATUS[0][0]:
+            messages.error(
+                request,
+                _("All of the Participations presence "
+                  "status must be initialised.")
+            )
+            return
+
+        # The status is present so we compute it in the report
+        if obj.presence_status == models.Participation.PRESENCE_STATUS[2][0]:
+            min_duration = int(obj.duration.total_seconds()/60)
+
+            if obj.user.pk in data_dict:
+                data_dict[obj.user.pk]['total_time'] += min_duration
+            else:
+                dataline = {
+                    'first_name': obj.user.first_name,
+                    'last_name': obj.user.last_name,
+                    'email': obj.user.email,
+                    'total_time': min_duration,
+                }
+                data_dict[obj.user.pk] = dataline
+
+    response = HttpResponse(content_type='text/csv')
+    response['Content-Disposition'] = \
+        'attachment; filename="%s.csv"' % cycle.name
+
+    writer = csv.writer(response)
+
+    writer.writerow(['first_name', 'last_name', 'email', 'total_time'])
+
+    for key, value in data_dict.items():
+        writer.writerow([
+            value['first_name'],
+            value['last_name'],
+            value['email'],
+            value['total_time']
+        ])
+
+    return response
```

Comment:
Can we create a generic functions to create the data export and just use this function to call the generic function ?
So when we will want to use it through the API it will be easy.

Reply:
I have made a function in Cycle to compile the data in a dict and then the admin receive it an manage it to make a csv